core/src/main/scala/org/apache/spark/config/ConfigEntry.scala

```diff
@@ -0,0 +1,247 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.config
+
+import java.util.concurrent.TimeUnit
+
+import org.apache.spark.SparkConf
+import org.apache.spark.network.util.{ByteUnit, JavaUtils}
+
+/**
+ * An entry contains all meta information for a configuration.
+ *
+ * @param key the key for the configuration
+ * @param defaultValue the default value for the configuration
+ * @param valueConverter how to convert a string to the value. It should throw an exception if the
+ *                       string does not have the required format.
+ * @param stringConverter how to convert a value to a string that the user can use it as a valid
+ *                        string value. It's usually `toString`. But sometimes, a custom converter
+ *                        is necessary. E.g., if T is List[String], `a, b, c` is better than
+ *                        `List(a, b, c)`.
+ * @param doc the document for the configuration
+ * @param isPublic if this configuration is public to the user. If it's `false`, this
+ *                 configuration is only used internally and we should not expose it to the user.
+ * @tparam T the value type
+ */
+private[spark] class ConfigEntry[T] (
+    val key: String,
+    val defaultValue: Option[T],
+    val valueConverter: String => T,
+    val stringConverter: T => String,
+    val doc: String,
+    val isPublic: Boolean) {
+
+  def defaultValueString: String = defaultValue.map(stringConverter).getOrElse("<undefined>")
+
+  /**
+   * Returns a new ConfigEntry that wraps the value in an Option, for config entries that do
+   * not require a value.
+   */
+  def optional: OptionalConfigEntry[T] = {
+    require(!defaultValue.isDefined, s"$this has a default value, cannot be optional.")
+    new OptionalConfigEntry(key, valueConverter, stringConverter, doc, isPublic)
```

Comment:
I thought about this one quite a bit. My understanding is that this exists because we have configs without default values and we just want to get the value as an option. However, in my mind all configs are optional to begin with (maybe except for `spark.master` etc.) but an "optional config" here means something different.

I think a more intuitive way might be just to add a `conf.getOption(configEntry: ConfigEntry[T])` that doesn't try to get the default value. If it's set, we get `Some(x)`, else if there's a default value we get `Some(y)`, otherwise `None`. This has three main benefits that I can think of:
- you don't need to add `.optional` after most configs
- it's clear that the return type is an option. Today `conf.get` sometimes returns an `Option`, which is weird to me.
- you don't really need the `readFrom` abstraction anymore (assuming we find another way to express fallback)

What do you think?